### PR TITLE
mtd/nand: Return -EUCLEAN when the bit error happen but fixed by ecc

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary

follow how Linux report the correction information from ecc:
http://www.infradead.org/pipermail/linux-mtd/2012-March/040305.html
since this information is very useful to file system(e.g. yaffs, ubi/buifs)
which is specially designed for nand flash.

## Impact

MTD NAND driver

## Testing

internal yaffs

### CODE REVIEW

File: drivers/mtd/mtd_nand.c [hunk 1]
```diff
@@ -600,7 +624,11 @@ static ssize_t nand_bread(FAR struct mtd_dev_s *dev, off_t startpage,
...
       ret = nand_readpage(nand, block, page, buffer);
-      if (ret < 0)
+      if (ret == -EUCLEAN)
+        {
+          fixedecc = true;
+        }
+      else if (ret < 0)
         {
           ferr("ERROR: nand_readpage failed block=%ld page=%d: %d\n",
                (long)block, page, ret);
```
Comment (pkarashchenko) [inline drivers/mtd/mtd_nand.c:634]:
```suggestion
          ferr("ERROR: nand_readpage failed block=%" PRIdOFF " page=%d: %d\n",
               block, page, ret);
```

File: drivers/mtd/mtd_nandecc.c [hunk 2]
```diff
@@ -138,14 +138,19 @@ int nandecc_readpage(FAR struct nand_dev_s *nand, off_t block,
...
+  switch (ret)
     {
+      case HAMMING_SUCCESS:
+        return OK;
+
+      case HAMMING_ERROR_SINGLEBIT:
+        return -EUCLEAN;
+
+      default:
       ferr("ERROR: Block=%d page=%d Unrecoverable error: %d\n",
```
Comment (pkarashchenko) [inline drivers/mtd/mtd_nandecc.c:150]:
```suggestion
      ferr("ERROR: Block=%" PRIdOFF " page=%d Unrecoverable error: %d\n",
```

File: drivers/mtd/mtd_nand.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/mtd/mtd_nand.c:634]: Done.

File: drivers/mtd/mtd_nandecc.c (hunk 2 above)
Comment (xiaoxiang781216) [reply drivers/mtd/mtd_nandecc.c:150]: Done.

File: drivers/mtd/mtd_nand.c [hunk 3]
```diff
@@ -110,6 +111,48 @@ static int     nand_ioctl(FAR struct mtd_dev_s *dev, int cmd,
...
+  scheme = nandmodel_getscheme(model);
+
+  /* Try to mark the block as BAD */
+
+  memset(spare, 0xff, CONFIG_MTD_NAND_MAXPAGESPARESIZE);
+  nandscheme_writebadblockmarker(scheme, spare, NAND_BLOCKSTATUS_BAD);
+  ret = NAND_WRITEPAGE(nand->raw, block, 0, 0, spare);
+  if (ret < 0)
+    {
+      ferr("ERROR: Failed bo marke block %" PRIdOFF " as BAD\n", block);
```
Comment (pkarashchenko) [inline drivers/mtd/mtd_nand.c:150]: Add `#include <inttypes.h>`

Comment (xiaoxiang781216) [reply drivers/mtd/mtd_nand.c:151]: Done.

**Review by pkarashchenko — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

**Pull request merged**